sbin/dmesg/dmesg.8
.\"	$NetBSD: dmesg.8,v 1.21.2.1 2019/06/10 22:05:32 christos Exp $
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)dmesg.8	8.1 (Berkeley) 6/5/93
.\"
.Dd October 30, 2018
.Dt DMESG 8
.Os
.Sh NAME
.Nm dmesg
.Nd display the system message buffer
.Sh SYNOPSIS
.Nm
.Op Fl dTt
.Op Fl M Ar core
.Op Fl N Ar system
.Sh DESCRIPTION
.Nm
displays the contents of the system message buffer.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Show the timestamp deltas.
Used together with
.Fl t
only the deltas are shown.
.It Fl M
Extract values associated with the name list from the specified core
instead of the default ``/dev/mem''.
.It Fl N
Extract the name list from the specified system instead of the default
``/netbsd''.
.It Fl T
Format uptime timestamps in a human readable form (using
.Xr ctime 3 )
using output suitable for the local locale as set in the environment.
Repeating this option prints the uptime in ISO 8601 duration form,
giving the duration since boot, in hours, minutes, and seconds (to
millisecond resolution).
A third occurrence causes the duration to always be represented
to millisecond precision, even where that means trailing zeroes
appear.
.It Fl t
Quiet printing, don't print timestamps.
.El
.Pp
The system message buffer is a circular buffer of a fixed size.
If the buffer has been filled, the first line of the
.Nm
output may not be complete.
The size of the message buffer is configurable at compile-time on
most systems with the
.Dv MSGBUFSIZE
kernel option.
Look for
.Dv MSGBUFSIZE
in
.Xr options 4
for details.
.Sh FILES
.Bl -tag -width /var/run/dmesg.boot -compact
.It Pa /var/run/dmesg.boot
copy of dmesg at the time of last boot.
.El
.Sh SEE ALSO
.Xr options 4 ,
.Xr syslogd 8
.Sh HISTORY
The
.Nm
command appeared in
.Bx 3.0 .
.Sh BUGS
The
.Fl T
option will report nonsense when displaying lines from
the message buffer that were not added by the current
running kernel.
.Pp
When
.Fl TT
is used, the duration is always given with maximum units of hours,
even when the number of hours is in the hundreds, thousands, or more.
This is because converting hours to days, over periods when
.Dq time skips
occur, such as summer time beginning or ending, is not trivial.
A duration of 26 hours might be 1D3H or 1D1H at such events,
rather than the usual 1D2H,
and when a time zone alters its offset,
even more complex calculations are needed.
None of those calculations are done
.Pq even to account for sub-hour time zone shifts ,
the duration indicated is always calculated by simple division of
seconds by 60 to produce minutes, and again to produce hours.
Most of the time\ [!] this is correct.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2405.2.1 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 8.0 to NetBSD 9.0:
	byacc: Update to 20170430. [christos 20170605]
	vax: Add support for VAXstation 4000 TURBOchannel. [flxd 20170609]
	wsbell(4): Added console bell support for all speaker devices, not
		only those attached at pcppi. [nat 20170612]
	bind: Import version 9.10.5-P1. [christos 20170615]
	dts: Import dts files from Linux 4.11.5. [jmcneill 20170615]
	amd64: Support 16TB of PA, and 32TB of kernel VA on the architecture
		side. [maxv 20170617]
	expat: Import 2.2.1 (security fixes). [christos 20170617]
	evbarm: Add support for Allwinner H3 SoC. [jmcneill 20170628]
	bind: Import version 9.10.5-P2. [christos 20170630]
	sunxiemac(4): Add support for Allwinner Gigabit Ethernet (EMAC).
		[jmcneill 20170701]
	fsdb(8): Add "saveblks" command. [mrg 20170804]
	wbsio(4): Add support for Winbond W83627DHG-P, W83627SF, W83627UHG,
		W83667HGB, W83687THF, W83697UG,	Nuvoton NCT5104D, NCT6775,
		NCT6779, NCT6791, NCT6792 and NCT6793. [msaitoh 20170707]
	evbarm: Add support for Allwinner A83T SoC. [jmcneill 20170707]
	MAKEDEV(8): Regenerate, documenting lua, dk, plcom, wmcom, hdmicec, mfi,
		nvme, nvme ns, tpm, and dtrace devices. [sevan 20170710]
	wbsio(4): Add support for NCT6795. [msaitoh 20170711]
	lm(4): Add support for NCT5174D, NCT6775F, NCT6779D and NCT679[1235]D.
		[msaitoh 20170711]
	evbarm: Add support for NVIDIA Tegra X1 SoC. [jmcneill 20170720]
	evbarm: Add support for Allwinner A64 SoC. [jmcneill 20170723]
	gzip(1): Add SIGINFO support. [mrg 20170804]
	i386: Remove vm86. [maxv 20170812]
	itesio(4): Add support IT8628E. [hauke 20170814]
	ichsmb(4): Add support for Intel C3000. [msaitoh 20170817]
	ismt(4): Add support for Intel C3000. [msaitoh 20170817]
	powerpc: Bump MAXTSIZ to 128MB for OEA based systems. [sevan 20170818]
	GMP/MPFR/MPC: Update to 6.1.2, 3.1.5 and 1.0.3. [mrg 20170821]
	ixg(4), ixv(4): Add C3000 support. Add bypass function support for
		bypass adapters. [msaitoh 20170830]
	file(1): Upgraded to 5.32. [christos 20170908]
	tcpdump(8): Import 4.9.2. [christos 20170908]
	veriexecgen(8): Drop support for MD5, SHA1, RMD160. [sevan 20170909]
	itesio(4): Add support IT8728GF and IT877[12]E [hauke 20170814]
	evbarm: Add support for Allwinner A13 and R8 SoCs. [jmcneill 20170825]
	veriexec(4): Drop support for VERIFIED_EXEC_FP_MD5,
		VERIFIED_EXEC_FP_SHA1, and VERIFIED_EXEC_FP_RMD160 options.
		[sevan 20170913]
	acpi(4): Updated ACPICA to 20170831. [christos 20170915]
	dtrace: Install dtruss, execsnoop, opensnoop, and procsystime from the
		DTrace Toolkit if DTrace support is enabled. [sevan 20170917]
	dhcpcd(8): Import dhcpcd-7.0.0-rc3. [roy 20171007]
	ahcisata(4), siisata(4): Add support for NCQ. [jdolecek 20171007]
	mvsata(4): Add support for NCQ, and enable ATAPI support.
		[jdolecek 20171007]
	OpenSSH: Imported 7.6. [christos 20171007]
	amd64: Add support for Kernel ASLR [maxv 20171010]
	evbarm: Add support for Allwinner H5 SoC. [jmcneill 20171011]
	tmux(1): Imported 2.6. [christos 20171012]
	amd64: Add support for SMAP. [maxv 20171017]
	amd64: Add support for USER_LDT. [maxv 20171021]
	tzdata Updated to 2017c. [kre 20171024]
	libc: Update to tzcode2017c. [christos 20171024]
	bwfm(4): Add driver for Broadcom BCM43xxx "FullMAC" USB wireless
		adapters. [jmcneill 20171025]
	kernhist(9): Reworked the export-via-sysctl code to use uintptr_t
		arguments, PR kern/52639.  Requires contemporaneous update
		of vmstat(1). [pgoyette 20171028]
	evbarm: Add support for NextThing GR8 SoC. [jmcneill 20171106]
	acpi(4): Updated ACPICA to 20171110. [christos 20171111]
	sunxinand(4): Add driver for Allwinner NAND Flash Controller.
		[jmcneill 20171113]
	gcc(1): Import GCC 5.5.  [mrg 20171113]
	libtre: Update to the latest git source. [rin 20171117]
	qemufwcfg(4): Add driver for QEMU Firmware Configuration device.
		[jmcneill 20171125]
	gdb(1): Updated to 8.0.1.  [christos 20171128]
	dhcpcd(8): Import dhcpcd-7.0.0-rc4. [roy 20171206]
	ihidev(4), ims(4): Added drivers for i2c HID mice. [bouyer 20171210]
	sparc: Fix time goes backwards. [mrg 20180112]
	lua: Applied 6th bugfix to Lua 5.3.4 from lua.org/bugs.html.
		[mbalmer 20171213]
	sstk: Remove the "stack section size change" syscall [kamil 20171219]
	sbrk(2): Remove the "data section size change" syscall [kamil 20171219]
	ovadvise: Remove the old vadvise syscall header [kamil 20171219]
	dhcpcd(8): Import dhcpcd-7.0.0 [roy 20180101]
	mdnsd(8), dns-sd(1), libdns_sd: import mDNSResponder 878.1
		[christos 20180114]
	cpuctl(8): Add cpuctl identify support for sparc and sparc64.
		[mrg 20180116]
	macppc: Enable Veriexec support in the GENERIC kernel [sevan 20180117]
	raid(4): Add support for 32-bit on 64-bit kernel.  [mrg 20180117]
	zoneinfo: Import tzdata2018c (skipping 2018a & 2018b). [kre 20180124]
	libc: Update to tzcode2018c. [christos 20180125]
	i386: Add support for SMAP. [maxv 20180128]
	dhcpcd(8): Import dhcpcd-7.0.1 [roy 20180129]
	gcc(1): Import GCC 6.4.  [mrg 20180201]
	openldap: Import 2.4.45. [christos 20180205]

sbin/dmesg/dmesg.c
/*	$NetBSD: dmesg.c,v 1.33.2.1 2019/06/10 22:05:32 christos Exp $	*/
/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1991, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)dmesg.c	8.1 (Berkeley) 6/5/93";
#else
__RCSID("$NetBSD: dmesg.c,v 1.33.2.1 2019/06/10 22:05:32 christos Exp $");
#endif
#endif /* not lint */

#include <sys/param.h>
#include <sys/msgbuf.h>
#include <sys/sysctl.h>

#include <err.h>
#include <ctype.h>
#include <fcntl.h>
#include <time.h>
#include <kvm.h>
#include <nlist.h>
#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <unistd.h>
#include <vis.h>

#ifndef SMALL
#include <langinfo.h>
#include <locale.h>

static struct nlist nl[] = {
#define	X_MSGBUF	0
	{ .n_name = "_msgbufp" },
	{ .n_name = NULL },
};

static const char *radix;

__dead static void	usage(void);

#define	KREAD(addr, var) \
	kvm_read(kd, addr, &var, sizeof(var)) != sizeof(var)

static const char *
fmtydhmsf(char *b, size_t l, intmax_t t, long nsec, int ht)
{
	intmax_t s, m, h;
	int z;
	int prec;
	size_t o;

	s = t % 60;
	t /= 60;

	m = t % 60;
	t /= 60;

	h = t;

	z = 0;
	o = 0;

#define APPENDFMT(fmt, ...)  \
    do { \
	    z = snprintf(b + o, l - o, fmt, __VA_ARGS__); \
	    if (z == -1) \
		    return b; \
	    o += (size_t)z; \
	    if (o >= l) \
		    return b; \
    } while (/*CONSTCOND*/0)

#define APPEND(a) \
    do if (a) \
	APPENDFMT("%jd%c", a, toupper((unsigned char)__STRING(a)[0])); \
    while (/*CONSTCOND*/0)
#define APPENDS(a, pr, ms) \
    APPENDFMT("%jd%s%.*ld%c", a, radix, pr, ms, \
	toupper((unsigned char)__STRING(a)[0]))

	APPENDFMT("%s", "P");
	APPENDFMT("%s", "T");
	APPEND(h);
	APPEND(m);
	if (nsec)
		nsec = (nsec + 500000) / 1000000;	/* now milliseconds */
	prec = 3;
	if (nsec && ht == 2) {
		while (prec > 0 && (nsec % 10) == 0)
			--prec, nsec /= 10;
	}
	if (nsec || ht > 2)
		APPENDS(s, prec, nsec);
	else
		APPEND(s);
	return b;
}

static void
pnsec(long nsec, long fsec, int scale)
{
	if (scale > 6)
		printf("%6.6ld", (nsec + 499) / 1000);
	else
		printf("%*.*ld%.*s", scale, scale, fsec, 6 - scale, "000000");
}
#endif

int
main(int argc, char *argv[])
{
	struct kern_msgbuf cur;
	int ch, newl, log, i;
	size_t tstamp, size;
	char *p, *bufdata;
	char buf[5];
#ifndef SMALL
	char tbuf[64];
	char *memf, *nlistf;
	struct timespec boottime;
	struct timespec lasttime;
	intmax_t sec;
	long nsec, fsec;
	int scale;
	int deltas, quiet, humantime;
	bool frac;

	static const int bmib[] = { CTL_KERN, KERN_BOOTTIME };
	size = sizeof(boottime);

	(void)setlocale(LC_ALL, "");
	radix = nl_langinfo(RADIXCHAR);
	if (radix == NULL)
		radix = ".";	/* could also select "," */

	boottime.tv_sec = 0;
	boottime.tv_nsec = 0;
	lasttime.tv_sec = 0;
	lasttime.tv_nsec = 0;
	deltas = quiet = humantime = 0;

	(void)sysctl(bmib, 2, &boottime, &size, NULL, 0);

	memf = nlistf = NULL;
	while ((ch = getopt(argc, argv, "dM:N:tT")) != -1)
		switch(ch) {
		case 'd':
			deltas = 1;
			break;
		case 'M':
			memf = optarg;
			break;
		case 'N':
			nlistf = optarg;
			break;
		case 't':
			quiet = 1;
			break;
		case 'T':
			humantime++;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;
	if (quiet && humantime)
		err(EXIT_FAILURE, "-t cannot be used with -T");

	if (memf == NULL) {
#endif
		static const int mmib[2] = { CTL_KERN, KERN_MSGBUF };

		if (sysctl(mmib, 2, NULL, &size, NULL, 0) == -1 ||
		    (bufdata = malloc(size)) == NULL ||
		    sysctl(mmib, 2, bufdata, &size, NULL, 0) == -1)
			err(1, "can't get msgbuf");

		/* make a dummy struct msgbuf for the display logic */
		cur.msg_bufx = 0;
		cur.msg_bufs = size;
#ifndef SMALL
	} else {
		kvm_t *kd;
		struct kern_msgbuf *bufp;

		/*
		 * Read in message buffer header and data, and do sanity
		 * checks.
		 */
		kd = kvm_open(nlistf, memf, NULL, O_RDONLY, "dmesg");
		if (kd == NULL)
			exit (1);
		if (kvm_nlist(kd, nl) == -1)
			errx(1, "kvm_nlist: %s", kvm_geterr(kd));
		if (nl[X_MSGBUF].n_type == 0)
			errx(1, "%s: msgbufp not found", nlistf ? nlistf :
			    "namelist");
		if (KREAD(nl[X_MSGBUF].n_value, bufp))
			errx(1, "kvm_read: %s (0x%lx)", kvm_geterr(kd),
			    nl[X_MSGBUF].n_value);
		if (kvm_read(kd, (long)bufp, &cur,
		    offsetof(struct kern_msgbuf, msg_bufc)) !=
		    offsetof(struct kern_msgbuf, msg_bufc))
			errx(1, "kvm_read: %s (0x%lx)", kvm_geterr(kd),
			    (unsigned long)bufp);
		if (cur.msg_magic != MSG_MAGIC)
			errx(1, "magic number incorrect");
		bufdata = malloc(cur.msg_bufs);
		if (bufdata == NULL)
			errx(1, "couldn't allocate space for buffer data");
		if (kvm_read(kd, (long)&bufp->msg_bufc, bufdata,
		    cur.msg_bufs) != cur.msg_bufs)
			errx(1, "kvm_read: %s", kvm_geterr(kd));
		kvm_close(kd);
		if (cur.msg_bufx >= cur.msg_bufs)
			cur.msg_bufx = 0;
	}
#endif

	/*
	 * The message buffer is circular; start at the write pointer
	 * (which points the oldest character), and go to the write
	 * pointer - 1 (which points the newest character).  I.e, loop
	 * over cur.msg_bufs times.  Unused area is skipped since it
	 * contains nul.
	 */
#ifndef SMALL
	frac = false;
	scale = 0;
#endif
	for (tstamp = 0, newl = 1, log = i = 0, p = bufdata + cur.msg_bufx;
	    i < cur.msg_bufs; i++, p++) {

#ifndef SMALL
		if (p == bufdata + cur.msg_bufs)
			p = bufdata;
#define ADDC(c)				\
    do {				\
	if (tstamp < sizeof(tbuf) - 1)	\
		tbuf[tstamp++] = (c);	\
	if (frac)			\
		scale++;		\
    } while (/*CONSTCOND*/0)
#else
#define ADDC(c)
#endif
		ch = *p;
		if (ch == '\0')
			continue;
		/* Skip "\n<.*>" syslog sequences. */
		/* Gather timestamp sequences */
		if (newl) {
#ifndef SMALL
			int j;
#endif

			switch (ch) {
#ifndef SMALL
			case '[':
				frac = false;
				scale = 0;
				ADDC(ch);
				continue;
#endif
			case '<':
				log = 1;
				continue;
			case '>':
				log = 0;
				continue;
#ifndef SMALL
			case ']':
				frac = false;
				ADDC(ch);
				ADDC('\0');
				tstamp = 0;
				sec = fsec = 0;
				switch (sscanf(tbuf, "[%jd.%ld]", &sec, &fsec)){
				case EOF:
				case 0:
					/*???*/
					continue;
				case 1:
					fsec = 0;
					break;
				case 2:
					break;
				default:
					/* Help */
					continue;
				}

				for (nsec = fsec, j = 9 - scale; --j >= 0; )
					nsec *= 10;
				if (!quiet || deltas)
					printf("[");
				if (humantime == 1) {
					time_t t;
					struct tm tm;

					t = boottime.tv_sec + sec;
					if (nsec + boottime.tv_nsec >=
					    ( 1L		/* 1 second */
						 * 1000L	/* ms */
						 * 1000L	/* us */
						 * 1000L	/* ns */ ))
							t++;

					if (localtime_r(&t, &tm) != NULL) {
						strftime(tbuf, sizeof(tbuf),
						    "%a %b %e %H:%M:%S %Z %Y",
						    &tm);
						printf("%s", tbuf);
					}
				} else if (humantime > 1) {
					const char *fp = fmtydhmsf(tbuf,
					    sizeof(tbuf), sec, fsec, humantime);
					if (fp) {
						printf("%s", fp);
					}
				} else if (!quiet) {
					printf(" %5jd%s", sec, radix);
					pnsec(nsec, fsec, scale);
				}
				if (deltas) {
					struct timespec nt = { sec, nsec };
					struct timespec dt;

					timespecsub(&nt, &lasttime, &dt);
					if (humantime || !quiet)
						printf(" ");
					printf("<% 4jd%s%6.6ld>",
					    (intmax_t)dt.tv_sec, radix,
					    (dt.tv_nsec+499) / 1000);
					lasttime = nt;
				}
				if (!quiet || deltas)
					printf("] ");
				continue;
#endif
			case ' ':
				if (!tstamp)
					continue;
				/*FALLTHROUGH*/
			default:
#ifndef SMALL
				if (tstamp) {
				    ADDC(ch);
				    if (ch == '.')
					frac = true;
				    continue;
				}
#endif
				if (log)
					continue;
				break;
			}
		}
		newl = ch == '\n';
		(void)vis(buf, ch, VIS_NOSLASH, 0);
#ifndef SMALL
		if (buf[1] == 0)
			(void)putchar(buf[0]);
		else
#endif
			(void)printf("%s", buf);
	}
	if (!newl)
		(void)putchar('\n');
	return EXIT_SUCCESS;
}

#ifndef SMALL
static void
usage(void)
{

	(void)fprintf(stderr, "Usage: %s [-dTt] [-M core] [-N system]\n",
		getprogname());
	exit(EXIT_FAILURE);
}
#endif
